### PR TITLE
Introduce `Value_Type.Null`, fix edge cases for `read_many`

### LINKED ISSUE #6281: Introduce a `Value_Type.Null`.
This is orthogonal to #5872.

Initially, I thought we can live without `Value_Type.Null`. Each of columns coming from the DB has some proper type. 

However, there are cases where this becomes problematic.

When I do `Table.new [["X", [Nothing]]]` - what type `X` should have? I guess for in-memory it's ok if it's mixed by default and we can cast it to any nullable type we want, and all will work OK.

However, there is the case of creating 'pure NULL' column in Expressions, both in-memory and in-DB.
Currently the behaviour is as follows:
```
t = table_builder [["X", [1, 2, 3]]]
t.compute "not(Nothing)" . should_fail_with Invalid_Value_Type
t.compute "Nothing + Nothing" . should_fail_with Illegal_Argument
t.compute "Nothing * Nothing" . should_fail_with Invalid_Value_Type
```

So any expression on nulls fails with Invalid Value Type, that's because the value is `Mixed` (in-memory) or `Unsupported_Data_Type "NULL"` (in-DB, at least in Postgres, I'm not even checking SQLite 🤕).

I'm not sure if that behaviour is right. It feels that if `t.at "X" + Nothing` works OK (if `X` is itself a numeric column), probably `[X] + Nothing` but also `Nothing + [X]` should work too.

The sane way to make it work that comes to my mind is only by adding a `Value_Type.Null` which these all-null columns can take. Such a special type can then be accepted by any `Value_Type.expect_*` operation, making these operations work.

- [ ] Add the `Null` type. It should not be mapped to-from SQL by default, I think.
- [ ] Make sure all `expect_*` functions also accept the `Null` type.
	- Should `is_*` answer True to it too? Figure out, initial thought is that probably not.
- [ ] Add tests for handling of these NULL columns in expressions.
- [ ] Update tests in `Aggregate_Column_Spec.enso`.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #6281

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

**Comment by radeusgd** [comment]
Changes affect Snowflake so I'm running the Extra suite:
- https://github.com/enso-org/enso/actions/runs/12145121839
- https://github.com/enso-org/enso/actions/runs/12158622104

I expect it to fail because it is failing on develop. Yet I want to see if there are no more failures than what is expected.

**Comment by radeusgd** [comment]
Followup to fix `Nothing * Nothing`: #11751 and `Nothing` in expressions in DB #11757.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso [hunk 1]
```diff
@@ -628,6 +626,9 @@ make_custom_cast column target_value_type type_mapping =
     result = Ref.new Nothing
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso:635]: I find this logic hard to follow. The return value would be either `Nothing` or the return value of `result.put expr` -- but isn't the the return value of `Ref.put` the old value?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Value_Type_Helpers.enso [hunk 2]
```diff
@@ -210,18 +210,29 @@ type Addition_Kind
     ## PRIVATE
     Text_Concat
 
+## Returns the operation kind based on types of the inputs.
+   If both inputs are Null, it is impossible to tell the kind, so Nothing is returned and the caller may decide what to do.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Value_Type_Helpers.enso:214]:
```suggestion
   If both inputs are Null, it is impossible to tell the kind, so Nothing is returned and the caller may decide what to do.
   Otherwise, the kinds must be the same `Addition_Kind`, and that kind is returned.
```

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso:635]: The return value of `put` is discarded.

The function returns `ref.get` at the very end.

I agree it is not ideal. This is a hack to get multiline if without adding a ton of indentation to the `else` branches. There's just too many branches here (and while sometimes too many branches is bad, here it is just intrinsic to this function) to make it readable.

Once we get multiline `if` (#6408) I will be more than happy to replace this with just:
```
if cond1 then
    ...
else if cond2 then
    ...
else
    Nothing
```

Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso:635]: Do you think I should rewrite it with increasing indentation?

I'd rather keep it as is as IMO it is readable once the reasons are explained.

**Comment by radeusgd** [comment]
> Followup to fix `Nothing * Nothing`: #11751 and `Nothing` in expressions in DB #11757.

I think we could merge this PR and then improve the missing bits from these 2 tickets in a follow-up PR. Unless there is any opposing views and we want to do it as a single change?

**Comment by radeusgd** [comment]
I've rebased on latest develop and noticed there are some test failures because of other changes. I will try to fix these soon, but it shouldn't change the shape of the overall changes that much.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso (hunk 1 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso:635]: I see, it makes more sense now.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso [hunk 3]
```diff
@@ -334,16 +360,19 @@ type Value_Type
...
         bucket_2 = find_bucket type_2
-        if bucket_1.is_nothing || bucket_2.is_nothing then False else
-            if (type_1 == Value_Type.Mixed) || (type_2 == Value_Type.Mixed) then True else
-                bucket_1 == bucket_2
+        either_type_is_not_ordered = bucket_1.is_nothing || bucket_2.is_nothing
+        if either_type_is_not_ordered then False else
+            comparable_with_everything = [Value_Type.Mixed, Value_Type.Null]
+            if (comparable_with_everything.contains type_1) || (comparable_with_everything.contains type_2) then True else
+                if bucket_1.is_nothing || bucket_2.is_nothing then False else
+                    bucket_1 == bucket_2
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Value_Type.enso:375]:
```suggestion
                bucket_1 == bucket_2
```
Since we know `either_type_is_not_ordered` is false here.

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java [hunk 4]
```diff
@@ -29,6 +31,10 @@ public ColumnStorage apply(
       return boolStorage.makeNegated();
     }
 
+    if (storage.getType() instanceof NullType) {
+      return new NullStorage(Math.toIntExact(storage.getSize()));
```
Comment (GregoryTravis) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java:35]: Is there a reason getSize() can return `long` but `NullStorage` can only accept `int`? Is an exception here possible?

**Review by GregoryTravis — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java (hunk 4 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java:35]: `getSize` is the Enso-facing one so it returns `long` because that is the default storage for integers in Enso.

All of our storages take `int` size because most of them are backed by arrays (Null is inded an exception) so they cannot hold more than `int` can index.

Thanks for pointing this out, making me rethink this - `long getSize()` is just wrapper for `int size()` that 'translates' to the more directly Enso-friendly `long` value, but I should have just used `size()` here and avoided the conversion.

**Review by AdRiley — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java (hunk 4 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/unary/NotOperation.java:35]: Hmm after further checking - this function gets `ColumnStorage` which only has the `long` returning one in its interface...

I guess I'll just keep the `toIntExact` conversion. It shouldn't happen as long as our storages are `int` based. Once they aren't - the other code will need to be updated anyway. Does that seem ok?

File: std-bits/table/src/main/java/org/enso/table/data/table/Column.java [hunk 5]
```diff
@@ -158,27 +157,7 @@ public static Column fromItemsNoDateConversion(
...
-    StorageType storageType = StorageType.forBoxedItem(converted);
-    Builder builder = Builder.getForType(storageType, repeat, problemAggregator);
-    Context context = Context.getCurrent();
-    for (int i = 0; i < repeat; i++) {
-      builder.append(converted);
-      context.safepoint();
-    }
-
-    return new Column(name, builder.seal());
+    return new Column(name, Storage.fromRepeatedItem(item, repeat, problemAggregator));
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/Column.java:160]: @jdunkerley I've allowed myself to overwrite some of your cleanups from the other PR in this particular method with a fully custom `Storage::fromRepeatedItem`.

It also relies on `Builder.getForType` for the generic case but also has 'specializations' that create a more compact storage directly (`NullStorage` and `LongConstantStorage` that are constructed in O(1) instead of O(N)).

I think we do want this 'optimization'. In the future we can even add more `ConstantStorage` kinds. 

The major question is if `Storage` class is the right place for this method. I imagine once we start adding more backends it may not necessarily be. But for now I think it is relatively good place. So I'd keep it here and move it when it makes sense.

Does that seem OK?

**Comment by radeusgd** [comment]
New Extra tests run after the new changes:
- ✅ https://github.com/enso-org/enso/actions/runs/12915770991

File: std-bits/table/src/main/java/org/enso/table/data/table/Column.java (hunk 5 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/data/table/Column.java:160]: Yes fine to do. Will rebase and cope :)

**Review by jdunkerley — APPROVED** [review]
